src/main/java/org/spongepowered/api/command/parameter/CommandContext.java

```diff
@@ -0,0 +1,163 @@
+     *
+     * @param key The key to look up
+     * @return whether there are any values present
+     */
+    boolean hasAny(Text key);
+
+    /**
+     * Returns whether this context has any value for the given argument key.
+     *
+     * @param key The key to look up
+     * @return whether there are any values present
+     */
+    boolean hasAny(String key);
+
+    /**
+     * Gets the value for the given key if the key has only one value.
+     *
+     * @param key the key to get
+     * @param <T> the expected type of the argument
+     * @return the argument
+     */
+    <T> Optional<T> getOne(Text key);
+
+    /**
+     * Gets the value for the given key if the key has only one value.
+     *
+     * @param key the key to get
+     * @param <T> the expected type of the argument
+     * @return the argument
+     */
+    <T> Optional<T> getOne(String key);
+
+    /**
+     * Gets the value for the given key if the key has only one value,
+     * otherwise, throws a {@link NoSuchElementException}.
+     *
+     * @param key the key to get
+     * @param <T> the expected type of the argument
+     * @return the argument
+     */
+    <T> T getOneUnchecked(Text key);
+
+    /**
+     * Gets the value for the given key if the key has only one value,
+     * otherwise, throws a {@link NoSuchElementException}.
+     *
+     * @param key the key to get
+     * @param <T> the expected type of the argument
+     * @return the argument
+     */
+    <T> T getOneUnchecked(String key);
+
+    /**
+     * Gets all values for the given argument. May return an empty list if no
+     * values are present.
+     *
+     * @param key The key to get values for
+     * @param <T> the type of value to get
+     * @return the collection of all values
+     */
+    <T> Collection<T> getAll(Text key);
+
+    /**
+     * Gets all values for the given argument. May return an empty list if no
+     * values are present.
+     *
+     * @param key The key to get values for
+     * @param <T> the type of value to get
+     * @return the collection of all values
+     */
+    <T> Collection<T> getAll(String key);
+
+    /**
+     * Adds a parsed object into the context, for use by commands.
+     *
+     * @param key The key to store the entry under.
+     * @param object The object to store.
+     */
+    void putEntry(Text key, Object object);
+
+    /**
+     * Adds a parsed object into the context, for use by commands.
+     *
+     * @param key The key to store the entry under.
+     * @param object The object to store.
+     */
+    void putEntry(String key, Object object);
+
+    /**
+     * Creates a snapshot of this context and returns a state object that can
+     * be used to restore the context to this state.
+     *
+     * <p>No assumptions should be made about the form of this state object,
+     * it is not defined in the API and may change at any time.</p>
+     *
+     * @return The state.
+     */
+    Object getState();
```

Comment:
We could. The idea behind returning an `Object` (from the original API) is that the consumer really shouldn't care about what's inside the object - just that it's something that can be used to restore state. I don't know what a snapshot would be used for, other than it would reduce error during development because you know you have the right snapshot. Keeping this as an `Object` just means that we be more liberal in the implementation on changing this if that's ever needed - but then, we can do that by just varying the contents of the snapshot object. 

So, I'm open to it either way.